lib/postprocessing/image.js

```diff
@@ -16,10 +16,163 @@ const gm = require("./gm-promisify");
+    "PNG",
+    "JPEG",
+    "JPEG",
+    "TIFF",
+    "TIFF",
+    "GIF",
+    "BMP"
+]);
+
+// minimal instructions which many workers can already handle themselves
+// const BASIC_INSTRUCTIONS = new Set([
+//     "fmt",
+//     "width",
+//     "height",
+//     "quality"
+// ]);
+
+// list of post processing instructions we handle
+const EXTRA_INSTRUCTIONS = new Set([
+    "crop",
+    "jpegSize",
+    "interlace",
+    "dpi",
+    "convertToDpi"
+]);
+
+function isSupportedOutputFormat(instructions) {
+    return Object.keys(SUPPORTED_OUTPUT_FMT).includes(instructions.fmt);
+}
+
+function isSupportedInputFormat(metadata) {
+    return SUPPORTED_INPUT_FMT.has(metadata.format);
+}
+
+function hasExtraInstructions(instructions) {
+    return Object.keys(instructions).some(key => EXTRA_INSTRUCTIONS.has(key));
+}
+
+async function getImageMetadata(file) {
+    try {
+        // [1x1+0+0] is a trick to avoid reading all pixels into memory, improving time for large images
+        const result = await gm(`${file}[1x1+0+0]`).write("json:-");
+        const json = JSON.parse(result.toString());
+
+        // newer imagemagick returns an array (for multi-page formats), older just one object (6.8.9-9 at least)
+        if (Array.isArray(json)) {
+            return json[0].image;
+        } else {
+            return json.image;
+        }
+    } catch (e) {
+        throw new Error(`Reading metadata from intermediate rendition failed: ${e.message}`);
+    }
+}
+
+function isTargetFormat(metadata, instructions) {
+    const format = SUPPORTED_OUTPUT_FMT[instructions.fmt];
+    if (format) {
+        return metadata.format === format;
+    }
+
+    return false;
+}
+
+function isExpectedSize(metadata, instructions) {
+    const dimension = metadata.pageGeometry;
+    if (instructions.width && instructions.height) {
+        return dimension.width === instructions.width && dimension.height === instructions.height;
+
+    } else if (instructions.width) {
+        // TODO: this is an optimistic check omitting validating height is in the right aspect ratio.
+        //       doing this would requiring reading the original source file dimensions which we
+        //       would like to avoid (extra time). for now we trust the worker
+        return dimension.width === instructions.width;
+
+    } else if (instructions.height) {
+        // TODO: this is an optimistic check omitting validating width is in the right aspect ratio.
+        //       doing this would requiring reading the original source file dimensions which we
+        //       would like to avoid (extra time). for now we trust the worker
+        return dimension.height === instructions.height;
+    }
+
+    // if no width or height is set, we assume nothing to be done,
+    // as rendition should be original size which we do not know here
+    return true;
+}
+
+function hasJpegQuality(metadata, instructions) {
+    if (metadata.format === "JPEG" && instructions.quality) {
```

Comment:
Is the format normalized? Or do we need to check for both JPG and JPEG?
Also, do we need to ensure metadata.format is upper case?

Reply:
The `format` returned by imagemagic are fixed and documented here: https://imagemagick.org/script/formats.php